### ISSUE TITLE
Review design on how we produce error summary

### ISSUE THREAD

**Issue by AndresTanasijczuk** [dmwm/CRABServer#4780]
Introduction and description of the current problem in CRAB 3.3.16

1) Each job has an exit code and exit message written in the FJR.
2) Until CRAB release 3.3.15, errors detected or happening in the post-processing stage were not added to the FJR file.
3) In the post-job we have a function prepareErrorSummary that creates/edits an error_summary.json file in the schedd containing the exit code and exit message for each job (each job id and each job retry).
4) This file is then read by the 'crab status' command to report the Error Summary to the user.
5) The FJR files read by the prepareErrorSummary function are the ones named "job_fjr.\<jobid\>.\<crab_retry\>.json". These files are created by the post-jobs of the corresponding jobid and crab_retry, when they star execution: they copy the "jobReport.json.\<jobid\>" file created by the job wrapper to "job_fjr.\<jobid\>.\<crab_retry\>.json".
6) Since we can only run 20 post-jobs at a time, and since the files "job_fjr.\<jobid\>.\<crab_retry\>.json" were not edited by the post-jobs, the prepareErrorSummary function was designed to read all existing "job_fjr.\<jobid\>.\<crab_retry\>.json" files of the task, independently of the job id or job retry count. This way the error summary could be reported quicker to users: as soon as the "job_fjr.\<jobid\>.\<crab_retry\>.json" file for a given (jobid, crab_retry) is available, any other post-job will read it and can insert the exit code and exit message to the error_summary.json file.
7) Also, since the "job_fjr.\<jobid\>.\<crab_retry\>.json" files were not edited by the post-jobs, if a failure occurs while retrieving the exit code or exit message from one of the "job_fjr.\<jobid\>.\<crab_retry\>.json" files or when loading the file, the prepareErrorSummary function inserts in the error_summary.json file exit code = -1 and exit message "Invalid Framework job report. bla, bla " or "Framework job report could be loaded, but no error message was found.".
8) Before trying to read a "job_fjr.\<jobid\>.\<crab_retry\>.json" file, the prepareErrorSummary function looks in the error_summary.json file if the exit code and exit message for that (jobid, crab_retry) was already inserted in the file. If it was, it will not change them, so it doesn't bother to even read the "job_fjr.\<jobid\>.\<crab_retry\>.json" file. This makes sense since, until CRAB 3.3.15, the "job_fjr.\<jobid\>.\<crab_retry\>.json" files once created are not further edited (and so whatever exit code and exit message was inserted in the error_summary.json file, they are terminal; they will not change).

9) In CRAB 3.3.16 we have made the following changes: 
a) Errors occurring in cmscp are written to the "jobReport.json.\<jobid\>" file and become automatically available in the error summary.
b) Errors occurring in the post-job are written to the "job_fjr.\<jobid\>.\<crab_retry\>.json" file, but in a different new section to not conflict with the exit code and exit message written by the job wrapper.
c) Errors like job killed in WN, which don't produce a valid FJR and are detected by the RetryJob (the RetryJob is called by the PostJob near the beginning of execution) are written into a newly created "job_fjr.\<jobid\>.\<crab_retry\>.json" file.
d) The prepareErrorSummary function was modified to write into the error_summary.json file the post-job exit code and exit message in case the (job) exit code is 0 and the post-job exit code != 0.

10) What was not foreseen in 9)b), 9)c) and 9)d) is that, if because of 6) and 7) another post-job has already (tried to) read the "job_fjr.\<jobid\>.\<crab_retry\>.json" file in the prepareErrorSummary function, the error_summary.json file will already contain an entry for this (jobid, crab_retry). And because of 8) this entry will not be overwritten with the new (and correct) exit code and exit message just added to the "job_fjr.\<jobid\>.\<crab_retry\>.json" file.

What is the change I propose?

Change the prepareErrorSummary function to not read all the "job_fjr.\<jobid\>.\<crab_retry\>.json" files, but only the one for the corresponding (jobid, crab_retry).
Pro: I think this is the more robust thing we can do in terms of avoiding future bugs. 
Con: This means that errors will not appear in the crab status Error Summary until the post-job for the given (jobid, crab_retry) finishes execution. The client will report something like "Exit code and exit message not (yet) found for N jobs." until the post-job for such a job finishes execution. 
Can we recover the property that error messages for already failed jobs are shown in the crab status Error Summary as soon as the post-job for that (jobid, crab_retry) starts to run? (This is a nice property IMO because post-jobs can take a long time until finish execution.) 
Yes, we can.
We can keep reading all the "job_fjr.\<jobid\>.\<crab_retry\>.json" files, but for alien (jobid, crab_retry) only write to the error_summary.json file in case the (job) exit code retrieved from the "job_fjr.\<jobid\>.\<crab_retry\>.json" file is not 0, because we are sure (I think) that in this case we don't care about any new information the post-job could write to the "job_fjr.\<jobid\>.\<crab_retry\>.json" file.
And can we have the error messages for already failed jobs shown in the crab status Error Summary even before the post-jobs run, given that post-job can be queued for long time until they start execution?
Well, for that we need the "job_fjr.\<jobid\>.\<crab_retry\>.json" files to be created by cmscp, which is not trivial, given that the crab_retry represents the number of time a post-job has started for a given jobid and this is of course not known in cmscp.

**Comment by belforte** [comment]
I almost fully like the change you propose. But it has too many "if", I'd like something more clear cut. I sort of have an idea, but have to leave (again, I know..). FWIW I will try to come back to this after dinner.

**Comment by belforte** [comment]
mostly the problem is that you try to guess for a given jobs whether postJob has run, will run, is running... we should bknow for sure from some clear source.
One should not "guess" if postJob may still resubmit or not.

Then things will be more clear.
- postJob all done: simple
- postJob runnning/pending : status is only intermediate, show job as
  - running if do not know any better
  - if know that it is finshed on WN but postJob has not completed yet, say so, possibly show exit code from last time the "grid" node ran. If you can tell that it ran and there's no fjr (out of memory, time...) then invent an ad hoc temporary code/name

As usual, there is not enough time during current validation to think about this, so your patch is OK, but I would like all of you to think about how to simplify.

**Comment by AndresTanasijczuk** [comment]
Actually, I forgot a very important point in my description (I'm sorry, I don't know how I could miss that important point) and is that:

* In the crab status Error Summary there only appear errors from jobs that are in status `failed`, which means post-job has finished. If post-job would be running/pending the status of the job would be `transferring`. So this is already as @belforte suggested in his first three bullets. 

In that sense the crab status Error Summary is doing what a very simple robust code should do IMHO: only show errors for jobs that are shown as failed. It could be extended in case job is being retried to include jobs that are running because they are being retried, and show error from last time job/postjob run. But not now.

Having said that, if we don't plan to change this property of the crab status Error Summary, I think we should simplify the prepareErrorSummary function in the post-job to not read all the "job_fjr.<jobid>.<crab_retry>.json" files, but only the one for the corresponding (jobid, crab_retry).
I hope I didn't miss anything else.

**Comment by belforte** [comment]
But how do you tell that a job is in |failed| vs. |transferring| ?
And how do you tell if it going to be retried due to some non-fatal
exit code ? I hope we do not label as |transferring| jobs waiting to
be resubmitted.

there is value in showing the exit code of jobs whose postJob has not run yet.
And if we simply have no exit code to show because job was evicted by
PeriodicRemove, so be it.

THe point is that "whatever signal is used to tell: postJob has run"
must be defined and cast in stone, can't be "now I look if exit code is 0
because I think it only can happen when postJob will not change things",
and somewhere else "I look at file names (or dates) to tell if postJob
will still do something or not" and somewhere else "I look at DAG status file" etc.

Even if all of those work.



On 04/29/2015 08:59 AM, Andres Tanasijczuk wrote:
> Actually, I forgot a very important point in my description (I'm sorry, I don't know how I could
> miss that important point) and is that:
>
>   * In the crab status Error Summary there only appear errors from jobs that are in status |failed|,
>     which means post-job has finished. If post-job would be running/pending the status of the job
>     would be |transferring|. So this is already as @belforte <https://github.com/belforte> suggested
>     in his first three bullets.
>
> In that sense the crab status Error Summary is doing what a very simple robust code should do IMHO:
> only show errors for jobs that are shown as failed. It could be extended in case job is being
> retried to include jobs that are running because they are being retried, and show error from last
> time job/postjob run. But not now.
>
> Having said that, if we don't plan to change this property of the crab status Error Summary, I think
> we should simplify the prepareErrorSummary function in the post-job to not read all the
> "job_fjr...json" files, but only the one for the corresponding (jobid, crab_retry).
> I hope I didn't miss anything else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmwm/CRABServer/issues/4780#issuecomment-97327712>.
>

-- 
Stefano Belforte - I.N.F.N.     tel   : +39 040 375-6261 (fax: 375-6258)
Area di Ricerca - Padriciano 99 tel mobile: +39 328 010 7327
34012 TRIESTE TS  - Italy       AIM: stefanobelforte

**Comment by mmascher** [comment]
Continuing the discussion from https://github.com/dmwm/CRABServer/pull/4778 (so we don't have some discussion on the ticket and some on the pull)

Discussin with Andres and Justas it turned out the proposed approach has a pitfall: before each postojob was reading the information about everybody, so concurrent updates of the error summary did not cause a problem. Now we are in a situation where what could happen is:

PJ A starts and read the error summary file
PJ B starts and read the error summary file
PJ A gets the error of Job A
PJ B gets the error of Job B
PJ A updates and write the error sumary file with its own report
PJ B updates and write the error sumary file (overwriting the report of PJ A)

A solution could be to use file locks in this case, instead of syncronizing writing temporary files and moving to the final one. It would work like this:

```python
import fcntl
file = open('error_summary.json', "r+")
fcntl.flock(file.fileno(), fcntl.LOCK_EX) # lock the file
#from now on only one process will be able to read/write the file, the others will be in "queue"
#read/write whatever we need
file.close() # unlocking the file
```

Better to use `with open` of course

**Comment by belforte** [comment]
good point. worth trying.
One comment:
problem is that we have all those DAG nodes which try to do "global" operations, like write a summary for the whole task. The need for this will decrease with new postJobs which can be restarted by HTC 3.8.5, as Marco explained me yesterday, but may not go away completely even for other use cases.

So I am wondering if we shouldn't think more seriously at adding one DAG node which is special, sort of only a preJob (or postJob) and simply lives as long as the DAG, watching all the job-related ones, and takes care of global task operations, like bulding reports. I do not know exactly how best to make DAG have a node that stops when "all the others are done", but it should be possible even if makes the overall control flow deviate from pure DAG. If the idea is liked maybe we can ask Brian for design suggestion. If we had that task-wide process, a few other things may become easier.

**Comment by mmascher** [comment]
I think we can add NOOP node for these kind of things (I hope I remember correctly the name, it means no operations). We were using it at some point but we removed because there was something not working in dagman that possibly is fixed now.

Then we can have two of such nodes:
1) One that "run in parallel" with the postjobs for reports
2) One that depends on all the jobs and only run when all the jobs are finished for close-out operations

**Comment by belforte** [comment]
NOOP is used to create a DAG node w/o HTCondor job, i.e. a pure preJob (or postJob) thing. so:
1) is a NOOP node
2) is a DAG FINAL node, which again could be NOOP. This is executed also if whole DAG fails or is killed.

The tricky point is to make 1) end when all other nodes are done. This is not part of DAG since in DAG nodes are fully independent of each other. One can created an external dependency by using the pre/postJob scripts to look at what other nodes do via files (logs, locks... read RunJobs.dag.nodes.log ) or whatever .... nothing really nice.
Another way (a bit cleaner):
- TW starts a two node DAG: the TaskMonitor (TM) and the JobsDag(JD). TM is NOOP
- JD is the current DAG, with a FINAL node.
- This FINAL node is NOOP and sends a termination message to TM.
In this way TM gets a termination message also if all job nodes die badly or are killed, w/o need to timeout etc.
Or maybe there is some other better way.

Or maybe there's some other way.

**Comment by mmascher** [comment]
That's my proposal using file locks: https://github.com/mmascher/CRABServer/commit/ecfc8f4ce63563e94071b3d15becc9d68ab2f430

@AndresTanasijczuk does it look ok? I made some tests but it would be better if also @juztas did some. I know he's on holiday, but still, maybe I'll wait him before merging.

**Comment by AndresTanasijczuk** [comment]
Looks ok to me. I can test it in my private VM submitting a similar task as Justas was submitting for the validation, that is, a task with 100 jobs where he set maxMemoryMB to some small value (Justas was using 20) so that all jobs are killed in WN because of excessive memory use. Then check that the error summary shows the 100 jobs have failed with that error.

**Comment by juztas** [comment]
@mmascher @AndresTanasijczuk give me few minutes to test it ;)

**Comment by juztas** [comment]
@mmascher output says everything:
```
No publication information (publication has been disabled in the CRAB configuration file)

Error Summary: (use --verboseErrors for details about the errors)

19 jobs failed with exit code -1

 1 jobs failed with exit code 50660

```

**Comment by AndresTanasijczuk** [comment]
Looking at the task Justas submitted for this test (http://glidemon.web.cern.ch/glidemon/jobs.php?taskid=1081606&showResubmits=1) I can see the same as happened to me: it is the RetryJob who is not catching that the jobs were removed due to memory use. Only in job 14 the reason was catch. And status is reporting that correctly:

$ crab status -d crab_lock-check-813582 --verboseErrors
CRAB project directory:		/afs/cern.ch/work/a/atanasi/CRAB3-tutorial/CMSSW_7_0_5/src/crab_lock-check-813582
Task name:			150501_195530:jbalcas_crab_lock-check-813582
Task status:			FAILED
Glidemon monitoring URL:	http://glidemon.web.cern.ch/glidemon/jobs.php?taskname=150501_195530%3Ajbalcas_crab_lock-check-813582
Dashboard monitoring URL:	http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=2&status=&site=&tid=150501_195530%3Ajbalcas_crab_lock-check-813582
Details:			failed        100.0% (20/20)

No publication information (publication has been disabled in the CRAB configuration file)

Error Summary:

19 jobs failed with exit code -1:

	19 jobs failed with following error message: (for example, job 1)

		The framework job report could be loaded, but no error message was found there.

 1 jobs failed with exit code 50660:

	 1 jobs failed with following error message: (for example, job 14)

		Not retrying job due to excessive memory use (job automatically killed on the worker node)

Have a look at https://twiki.cern.ch/twiki/bin/viewauth/CMSPublic/JobExitCodes for a description of the exit codes.
Log file is /afs/cern.ch/work/a/atanasi/CRAB3-tutorial/CMSSW_7_0_5/src/crab_lock-check-813582/crab.log


error_summary.json file here: http://vocms0114.cern.ch/mon/cms1761/150501_195530:jbalcas_crab_lock-check-813582/error_summary.json

[17:42:37][atanasi@vocms0114][/home/grid/cms1761/150501_195530:jbalcas_crab_lock-check-813582/]$ grep "Not retrying job due to excessive memory use" job_fjr.*.json
job_fjr.14.2.json:{"postjob": {"exitMsg": "", "exitCode": 2}, "exitMsg": "Not retrying job due to excessive memory use (job automatically killed on the worker node)", "exitCode": 50660}

So error_summary.json file is ok.

@juztas IIRC during the week when you were testing the new error summary code you had the same issue (that RetryJob was not detecting the correct remove reason) and I think you told me that it seemed there is a delay in making the reason available in the job ad (which is from where the RetryJob is reading it). And IIRC in later tests with the same task (you were resumitting the same task again and again) the remove reason was detected correctly by all the 100 jobs. Did you implemented a hack (e.g. added a sleep) in that original task? Or am I remembering wrong or confusing something? I am not saying we should implement you hack, but just want to understand how you were able to make all yours jobs detect the correct remove reason and maybe you could implement the same hack for this new task.

**Comment by juztas** [comment]
It was because @belforte explained better about locking job_log file and in that case we returned copy of job_log file (see https://github.com/dmwm/CRABServer/blob/6bcc8484f24bbb559f124ec462f0ea12660def53/src/python/TaskWorker/Actions/RetryJob.py#L19 which should not lock). I would need to take a look what Brian proposed and reimplement. But seems we are still missing this: https://github.com/dmwm/CRABServer/pull/4776 and with this it should work until we implement what Brian proposed.
@mmascher can you prepare tw rpm and I will install it for final validation?

**Comment by AndresTanasijczuk** [comment]
Ok. I don't understand about "locking job_log" nor about "what Brian proposed". But will not ask now.

But from what you said I understand that: you expect that once #4776 is included (because it is merged in master, but not included yet in current TW in testbed) the RetryJob will already be able to identify the correct remove reason. Is that correct?

**Comment by juztas** [comment]
Yes, tested on same task:
```
[jbalcas@lxplus0015 ~]$ crab status --verboseErrors
CRAB project directory:		/afs/cern.ch/user/j/jbalcas/Validation/Roles_GOOD/becms-role/CMSSW_7_0_6/src/twtest/crab_lock-check-813582
Task name:			150501_195530:jbalcas_crab_lock-check-813582
Task status:			FAILED
Glidemon monitoring URL:	http://glidemon.web.cern.ch/glidemon/jobs.php?taskname=150501_195530%3Ajbalcas_crab_lock-check-813582
Dashboard monitoring URL:	http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=2&status=&site=&tid=150501_195530%3Ajbalcas_crab_lock-check-813582
Details:			failed        100.0% (20/20)

No publication information (publication has been disabled in the CRAB configuration file)

Error Summary:

20 jobs failed with exit code 50660:

	20 jobs failed with following error message: (for example, job 1)

		Not retrying job due to excessive memory use (job automatically killed on the worker node)

Have a look at https://twiki.cern.ch/twiki/bin/viewauth/CMSPublic/JobExitCodes for a description of the exit codes.
Log file is /afs/cern.ch/user/j/jbalcas/Validation/Roles_GOOD/becms-role/CMSSW_7_0_6/src/twtest/crab_lock-check-813582/crab.log

```

**Comment by belforte** [comment]
So now we copy job log before using in condor_q , and it works, in spite of file locking being a red herring :-( 

On May 2, 2015 10:16 PM, Justas Balčas <notifications@github.com> wrote:Yes, tested on same task:

[jbalcas@lxplus0015 ~]$ crab status --verboseErrors
CRAB project directory:     /afs/cern.ch/user/j/jbalcas/Validation/Roles_GOOD/becms-role/CMSSW_7_0_6/src/twtest/crab_lock-check-813582
Task name:          150501_195530:jbalcas_crab_lock-check-813582
Task status:            FAILED
Glidemon monitoring URL:    http://glidemon.web.cern.ch/glidemon/jobs.php?taskname=150501_195530%3Ajbalcas_crab_lock-check-813582
Dashboard monitoring URL:   http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=2&status=&site=&tid=150501_195530%3Ajbalcas_crab_lock-check-813582
Details:            failed        100.0% (20/20)

No publication information (publication has been disabled in the CRAB configuration file)

Error Summary:

20 jobs failed with exit code 50660:

    20 jobs failed with following error message: (for example, job 1)

        Not retrying job due to excessive memory use (job automatically killed on the worker node)

Have a look at https://twiki.cern.ch/twiki/bin/viewauth/CMSPublic/JobExitCodes for a description of the exit codes.
Log file is /afs/cern.ch/user/j/jbalcas/Validation/Roles_GOOD/becms-role/CMSSW_7_0_6/src/twtest/crab_lock-check-813582/crab.log

**Comment by belforte** [comment]
maybe you want to close the original issue, but keep this open. We would still like to understand better the condor_q destructive interference and/or find the way to do w/o condor_q (aka: "we implement what Brian proposed" as Justas wrote). In a way we have two solutions:
1. condor_q on common log file, which fails, and we do not why
2. condor_q on copies of common log_file which at least works better, but lacking understanding of 1. we can not be sure it will always work
3. concurrent editing of common summary by multiple postJobs which is now cured via file locks but could probably be done in simpler way

**Comment by mmascher** [comment]
Hi Stefano, yes, let's keep this one open. It closed automatically because the commit had "Fixes ..."

**Comment by belforte** [comment]
assign to me the task to re-read, re-understand and decide if and what to do with this

**Comment by belforte** [comment]
relabel as BUG since that's the only thing left.
in the, either we do something wrong and have to fix it, or all is OK and close it.

**Comment by belforte** [comment]
since @emaszs is reworking the way to produce error_summary, and in the end condor_q appears not to fail in any relevant amount latetly. Let's close this.
